### DISCUSSION TITLE
Simulate copy command for a content item

### DISCUSSION THREAD

**Discussion by nicpitsch** [2sic/2sxc#2444] [Q&A] [answered]
I'm searching for a command to duplicate/copy an item for usage as a new item, as the Admin.. Data tab offers for host users:
![2021-06-08_14-06-58](https://user-images.githubusercontent.com/1652835/121189776-b5f8c480-c86a-11eb-9fd7-fac5001a415a.jpg)

The intention is to implement this with a Toolbar, to expose this functionality to non-admin users. But the [Commands an Parameters listing](https://docs.2sxc.org/js-code/commands/index.html#all-commands--parameters-todo---update) doesn't contain a *copy* command.

**Comment by iJungleboy** [comment, marked as answer]
A plain copy would also be a bit strange, because it's unclear what would happen with the new copied item. 

What the Toolbar-UI does expose is the replace-dialog, which then let's you create a copy of something existing. The same should also work for add-existing. I think this is what you're looking for.

**Reply by nicpitsch** [reply]
Thanks for the hints. It ended with `new` and `prefill`:

```c#
<div class="xxxsc-element">
    @Edit.Toolbar(actions: "new", contentType: "Content", prefill: new { 
        Title = "[KOPIE..] " + item.Title, 
        //DateFrom = item.DateFrom, 
        //Type_e = item.Type_e[0],
        //Category_e = item.Category_e,
        Teaser = item.Teaser,
        Body = item.Body,
        Images = item.Images,
        ImagesInRow = item.ImagesInRow,
        Docs = item.Docs,
        Url = item.Url,
        UrlKey = item.UrlKey + "_" + System.DateTime.Now.ToString("mmss"),
        Body2 = item.Body2
    }, settings: new { hover = "left" })
</div> 
```

What remains unsolved is the more complex properties of Data Type *Entity* like 'Type_e', 'Category_e' and the Images (references) which I wasn't able to copy into the now Item.

**Reply by iJungleboy** [reply]
ok. note that we probably never tried prefill with very long data, this could run into bumps. 

Categories: you should prefill an array or CSV of IDs or GUIDs, not sure any more which one, but it's possible.

Image: you probably mean the orignal value like `file:72`. You can get that on the DynamicEntity using Get("fieldname", ...) - there are some params which prevent automatic conversions. https://docs.2sxc.org/api/dot-net/ToSic.Sxc.Data.IDynamicEntity.html#ToSic_Sxc_Data_IDynamicEntity_Get_System_String_System_String_System_String_System_Boolean_

**Reply by nicpitsch** [reply]
Thanks, that helped me to make copy of field content of data type 'Entity' functional (my Type_e and Category_e):

```c#
<div class="xxxsc-element">
    @Edit.Toolbar(actions: "new", contentType: "Content", prefill: new { 
        Title = "[KOPIE..] " + item.Title, 
        //DateFrom = item.DateFrom, 
        DateTo = item.DateTo,
        Type_e = new string[] { String.Join(",", ((IEnumerable<dynamic>)item.Type_e).Select(a => AsDynamic(a).EntityGuid)) },
        Category_e = new string[] { String.Join(",", ((IEnumerable<dynamic>)item.Category_e).Select(a => AsDynamic(a).EntityGuid)) },
        Teaser = item.Teaser,
        Body = item.Body,
        // Images = item.Images,  //  not functional
        ImagesInRow = item.ImagesInRow,
        Docs = item.Docs,
        Url = item.Url,
        UrlKey = item.UrlKey + "_" + System.DateTime.Now.ToString("mmss"),
        Body2 = item.Body2
    }, settings: new { hover = "left" })
</div>
```

Now I'm stucked with 'Images'. The Images field is data type Link.. 'Library of files'. I use it for images with ADAM. I tried such things like:

* `@AsDynamic(Content).Get("Images", convertLinks: false)`.
* `@AsDynamic(item.Images).GetType().GetProperties()[0]`.
* `@AsAdam(item, "Images").Files.GetType().GetProperties()`.

But I'm stuck with that.

**Reply by iJungleboy** [reply]
I think the `@` is your problem - in Razor that means the data is output to the browser. I think your first example without @ will work :)
```
Images = item.Get("Images", convertLinks: false),
```
...assuming it's `Images` - it's probably `Image`

...unless it's a library of items. That simply cannot be quick-copied, as that's just stored in a special folder, the entity doesn't have a reference to it because the folder is automatically determined by the entity-guid.

**Reply by nicpitsch** [reply]
> ...unless it's a library of items. That simply cannot be quick-copied, as that's just stored in a special folder, the entity doesn't have a reference to it because the folder is automatically determined by the entity-guid.

&rarr;  `Images` is a *library of items*. So the journey ends here. But I'm quite happy with the result above.

> I think the `@` is your problem - in Razor that means the data is output to the browser. I think your first example without @ will work :)

Sorry, the context is missing, the expressions were used within tags (<small>...).
